src/audio/mod.rs

```diff
@@ -0,0 +1,113 @@
+use crate::{Pipeline, Result};
+use anyhow::anyhow;
+use cpal::{
+    traits::{DeviceTrait, HostTrait, StreamTrait},
+    Device,
+};
+use ringbuf::{Consumer, Producer, RingBuffer};
+use std::time::Duration;
+
+const LATENCY: f32 = 20.0;
+
+pub fn run(mut pipeline: Pipeline) -> Result<()> {
+    let (input_device, output_device) = devices()?;
+
+    let config: cpal::StreamConfig = input_device.default_input_config()?.into();
+
+    let latency_frames = (LATENCY / 1_000.0) * config.sample_rate.0 as f32;
+    let latency_samples = latency_frames as usize * config.channels as usize;
+
+    let ring = RingBuffer::new(latency_samples * 2);
+    let (mut producer, mut consumer) = ring.split();
+
+    for _ in 0..latency_samples {
+        producer.push(0.0).unwrap();
+    }
+
+    let input_data_fn = move |data: &[f32], _: &cpal::InputCallbackInfo| {
+        let output_fell_behind = write_frame(&mut producer, data);
+
+        if output_fell_behind {
+            eprintln!("output stream fell behind: try increasing latency");
+        }
+    };
+
+    let output_data_fn = move |output: &mut [f32], _: &cpal::OutputCallbackInfo| {
+        let (frame, input_fell_behind) = read_frame(&mut consumer, output.len());
+        pipeline.process(&frame, output);
+
+        if input_fell_behind {
+            eprintln!("input stream fell behind: try increasing latency");
+        }
+    };
+
+    println!(
+        "Attempting to build both streams with f32 samples and {:?}.",
+        config
+    );
+    let input_stream = input_device.build_input_stream(&config, input_data_fn, handle_error)?;
+    let output_stream = output_device.build_output_stream(&config, output_data_fn, handle_error)?;
+    println!("Successfully built streams.");
+
+    println!(
+        "Starting the input and output streams with {} milliseconds of latency.",
+        LATENCY
+    );
+    input_stream.play()?;
+    output_stream.play()?;
+
+    std::thread::sleep(Duration::from_secs(u64::MAX));
+
+    drop(input_stream);
+    drop(output_stream);
```

Comment:
I'm not sure why these need to be explicitly dropped (this is taken from an example). Usually that'd happen automatically as they go out of scope.